### PR TITLE
Added Utilization Graphs to Service Details

### PR DESCRIPTION
**Pull request by chalettu**
PV https://www.pivotaltracker.com/story/show/140269955

### CODE REVIEW

**Comment by chriskacerguis** [comment]
@chalettu can you please attach a screenshot so we can get that looked at to make sure we are on the right track?

**Comment by chalettu** [comment]
demo v2

**Comment by chalettu** [comment]
demo graphs
Demo if a particular graph has no data

**Comment by chriskacerguis** [comment]
@ManageIQ/core-ux how is that looking?

**Comment by chalettu** [comment]
@miq-bot add_label services
@miq-bot add_label fine/no

**Comment by serenamarie125** [comment]
This should match what we have in PF ... take a look here: http://www.patternfly.org/pattern-library/cards/utilization-trend-card/#/code/angular

screen shot 2017-05-08 at 1 40 10 pm

**Comment by chriskacerguis** [comment]
@serenamarie125 are you talking about the trend line at the bottom?

**Comment by serenamarie125** [comment]
@chriskacerguis nope, the design doesn't have the trend info.  I'm talking about when there is no data available.  @chalettu has a mock of that above, sorry if it wasn't explicit

**Comment by chriskacerguis** [comment]
ahhh...ok, cool.

**Comment by chalettu** [comment]
demo v3
@serenamarie125 , I updated the text and also the placement of the messaging below the subtitle line.   Does this look better?

**Comment by serenamarie125** [comment]
@chalettu can you make sure the horizontal lines are aligned ( under CPU, Memory & Storage ) regardless if there is data

**Comment by chalettu** [comment]
demo v4
Updated to make sure the line remains even if no data exists

**Comment by chalettu** [comment]
@miq-bot add_label my services / details

File: client/app/services/usage-graphs/usage-graphs.html [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                <div pf-donut-pct-chart ng-if="vm.cpuDataExists" config="vm.cpuChart.config" data="vm.cpuChart.data" center-label="vm.cpuChart.label"></div>
+    </div>
+    <div class="col-md-4 text-center">
+        <div>
+            <div class="metric-title" translate>Memory</div>
+            <div class="metric-details-container">
+                <div ng-if="vm.memoryDataExists">
+                    <span class="metric-number">{{vm.availableMemory}}</span>
+                    <span class="metric-details">
+                        <p translate>gb available</p>
```
Comment (serenamarie125) [inline client/app/services/usage-graphs/usage-graphs.html]: Need to follow PatternFly standards.  "GB" should be in all caps.  "Available" should have a capital "A".

File: client/app/services/usage-graphs/usage-graphs.html [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        <div pf-donut-pct-chart ng-if="vm.memoryDataExists" config="vm.memoryChart.config" data="vm.memoryChart.data" center-label="vm.memoryChart.label"></div>
+    </div>
+    <div class="col-md-4 text-center">
+        <div>
+            <div class="metric-title" translate>Storage</div>
+            <div class="metric-details-container">
+                <div ng-show="vm.storageDataExists">
+                    <span class="metric-number">{{vm.availableStorage}}</span>
+                    <span class="metric-details">
+                            <p translate>gb available</p>
```
Comment (serenamarie125) [inline client/app/services/usage-graphs/usage-graphs.html]: change "gb" to "GB" and "available" to "Available"

File: client/app/services/service-details/service-details.component.js [hunk 3]
```diff
@@ -66,7 +68,88 @@ function ComponentController($stateParams, $state, $window, CollectionsApi, Even
...
+  }
+
+  function getChartConfigs() {
+    const allocatedStorage = (vm.service.aggregate_all_vm_disk_space_allocated / 1073741824).toFixed(2); // convert bytes to gb
+    const usedStorage = (vm.service.aggregate_all_vm_disk_space_used / 1073741824).toFixed(2); // convert bytes to gb
+    let usedMemory = 0;
+    let usedCPU = 0;
+    let totalCPU = 0;
+    const allocatedMemory = vm.service.aggregate_all_vm_memory / 1024; // this metric is in mb
+    angular.forEach(vm.service.vms, (instance) => {
```
Comment (AllenBW) [inline client/app/services/service-details/service-details.component.js]: recommend using `vm.services.vms.forEach(..`

File: client/app/services/service-details/service-details.html [hunk 4]
```diff
@@ -7,8 +7,24 @@
         <li class="active">{{vm.title}}</li>
       </ol>
   </span>
+  <span class="service-details-toolbar-buttons">
```
Comment (AllenBW) [inline client/app/services/service-details/service-details.html]: These two icons don't appear to line up (from ss in pr) additionally, do they indicate the active view? From this [mockup](https://github.com/ManageIQ/manageiq-design/blob/master/UX/ui-service/Catalogs/design.md) the active view type is indicated by the breadcrumbs toolbar icon changing to that blueish color when selected.  For references on how this is done, take a look at some of that catalog code we had prior to yanking out the list view.

File: client/app/services/service-details/service-details.component.js [hunk 5]
```diff
@@ -30,10 +30,10 @@ function ComponentController($stateParams, $state, $window, CollectionsApi, Even
       availableTags: [],
       credential: {},
       listActions: [],
-
       // Functions
       hasCustomButtons: hasCustomButtons,
       disableStopButton: disableStopButton,
+      displayedContent: 'details',
```
Comment (chalettu) [inline client/app/services/service-details/service-details.component.js]: This doesnt need that, this merely is a control for which active screen of information someone is looking at.

File: client/app.js [hunk 6]
```diff
@@ -12,6 +12,8 @@ window.moment = require('moment');
 window.sprintf = require('sprintf-js').sprintf;
 
 // Vendor libraries, order matters
+window.c3 = require('c3/c3.js');
```
Comment (chalettu) [inline client/app.js]: Patternfly expects C3 to be global like this.  I will look into moving it into our global file though

Comment (AllenBW) [inline client/app.js]: ` +require('c3); ` doesn't cut mustard?

Comment (chalettu) [inline client/app.js]: nope I tried it and this was the only way to get it working.

File: client/app/services/service-details/service-details.component.js (hunk 5 above)
Comment (chalettu) [inline client/app/services/service-details/service-details.component.js]: I took a look at how catalogs had implemented something similar and I renamed the variable and implemented the same pattern.

File: client/app/services/service-details/service-details.component.js [hunk 7]
```diff
@@ -66,7 +68,88 @@ function ComponentController($stateParams, $state, $window, CollectionsApi, Even
...
+    let totalValue = 0;
+
+    if (angular.isDefined(used)) {
+      usedValue = used;
+      totalValue = total;
+    }
+
+    return {
+      config: {
+        units: 'MHz',
```
Comment (chalettu) [inline client/app/services/service-details/service-details.component.js]: Done

File: client/app/services/service-details/service-details.component.js (hunk 3 above)
Comment (chalettu) [inline client/app/services/service-details/service-details.component.js]: Done

File: client/app/services/service-details/service-details.html (hunk 4 above)
Comment (chalettu) [inline client/app/services/service-details/service-details.html]: I tweaked its positioning just a little.  I guess its just the spacing in the icon.

**Comment by chalettu** [comment]
graphs
Updated screenshot

File: client/app/services/usage-graphs/_usage-graphs.sass [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+.metric-details-container
+  height: 30px
+
+.metric-details
+  display: block
+  float: left
+  padding-left: 10px
+  padding-top: 3px
+
+  p
```
Comment (AllenBW) [inline client/app/services/usage-graphs/_usage-graphs.sass]: This makes me nervous, any way we can push this style off into a class and keep da `p` clean?

File: client/app/services/usage-graphs/usage-graphs.html [hunk 9]
```diff
@@ -0,0 +1,76 @@
+  <h2 translate>Utilization</h2>
+  <hr>
+  <div class="row">
+    <div class="col-md-4 text-center">
+                <div>
```
Comment (AllenBW) [inline client/app/services/usage-graphs/usage-graphs.html]: a div that's necessary (its got no classes... 🤔 )?  Less is more when possible, this looks to happen in a few places... maybe the question is, can we get rid of the classless div?

File: client/app/services/usage-graphs/usage-graphs.html [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+                    </span>
+                </div>
+            </div>
+            <hr>
+            <div ng-if="!vm.memoryDataExists">
+                <div>
+                    <i class="pf pficon-info"></i>
+                </div>
+                <div translate>
+                No data available
```
Comment (AllenBW) [inline client/app/services/usage-graphs/usage-graphs.html]: Maybe not for this pr, but something we might want to look into doing is create a no data available component, i gotta feeling that might have uses in more places than here...

File: client/app/services/usage-graphs/_usage-graphs.sass (hunk 8 above)
Comment (chalettu) [inline client/app/services/usage-graphs/_usage-graphs.sass]: This is indented to where its only p's contained inside of elements of the .metric-details class.   Its scoped so its not a global p change.  Just for paragraphs inside .metric-details

**Comment by chriskacerguis** [comment]
@chalettu can you please address @AllenBW comments?

**Comment by AllenBW** [comment]
I'll approve, their nitpicky things anyways...

**Pull request merged**